shared/Microsoft.Extensions.ObjectMethodExecutor.Sources/ObjectMethodExecutor.cs

```diff
@@ -0,0 +1,453 @@
+            var onCompletedFunc = Expression.Lambda<Action<object, Action>>(
+                Expression.Call(
+                    Expression.Convert(onCompletedParam1, awaiterType),
+                    awaiterOnCompletedMethod,
+                    onCompletedParam2),
+                onCompletedParam1,
+                onCompletedParam2).Compile();
+
+            Action<object, Action> unsafeOnCompletedFunc = null;
+            if (awaiterUnsafeOnCompletedMethod != null)
+            {
+                // var unsafeOnCompletedFunc = (object awaiter, Action continuation) => {
+                //     ((CustomAwaiterType)awaiter).UnsafeOnCompleted(continuation);
+                // };
+                var unsafeOnCompletedParam1 = Expression.Parameter(typeof(object), "awaiter");
+                var unsafeOnCompletedParam2 = Expression.Parameter(typeof(Action), "continuation");
+                unsafeOnCompletedFunc = Expression.Lambda<Action<object, Action>>(
+                    Expression.Call(
+                        Expression.Convert(unsafeOnCompletedParam1, awaiterType),
+                        awaiterUnsafeOnCompletedMethod,
+                        unsafeOnCompletedParam2),
+                    unsafeOnCompletedParam1,
+                    unsafeOnCompletedParam2).Compile();
+            }
+
+            // return new ObjectMethodExecutorAwaitable(
+            //     (object)getCustomAwaitable(),
+            //     getAwaiterFunc,
+            //     isCompletedFunc,
+            //     getResultFunc,
+            //     onCompletedFunc,
+            //     unsafeOnCompletedFunc);
+            var returnValueExpression = Expression.New(
+                _objectMethodExecutorAwaitableConstructor,
+                Expression.Convert(methodCall, typeof(object)),
+                Expression.Constant(getAwaiterFunc),
+                Expression.Constant(isCompletedFunc),
+                Expression.Constant(getResultFunc),
+                Expression.Constant(onCompletedFunc),
+                Expression.Constant(unsafeOnCompletedFunc, typeof(Action<object, Action>)));
+
+            var lambda = Expression.Lambda<MethodExecutorAsync>(returnValueExpression, targetParameter, parametersParameter);
+            return lambda.Compile();
+        }
+
+        private static bool IsAwaitable(
+            Type type,
+            out MethodInfo getAwaiterMethod,
+            out Type awaiterType,
+            out Type returnType,
+            out PropertyInfo isCompletedProperty,
+            out MethodInfo getResultMethod,
+            out MethodInfo onCompletedMethod,
+            out MethodInfo unsafeOnCompletedMethod)
+        {
+            // Based on Roslyn code: http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/Shared/Extensions/ISymbolExtensions.cs,db4d48ba694b9347
+
+            // Awaitable must have method matching "object GetAwaiter()"
+            getAwaiterMethod = type.GetRuntimeMethods().FirstOrDefault(m => 
```

Comment:
I dunno if using binding flags is better but this does only happen once per method so meh.